### PR TITLE
[Diagnostics] Offer 'is' replacement for unused 'if let' expression with optional operand

### PR DESCRIPTION
**Pull request by mininny**
```
let foo: Any? = 42 as Int

// current fix-it: Replace 'let bar = foo as? Int' with '(foo as? Int) != nil'
// better fix-it: Replace 'let bar = foo as? Int' with 'foo is Int'

if let bar = foo as? Int { 
    print("'foo' is of type 'Int'")
}
```
Currently, the above code does not produce the better fix-it for replacing the expression with `is` because the expression was wrapped inside `OptionalEvaluationExpr`.

This PR enables optional typed values to produce `is` replacement fixit. 

Resolves [SR-14646](https://bugs.swift.org/browse/SR-14646)

### CODE REVIEW

**Review by xedin — COMMENTED** [review]
Thank you for the contribution! Code makes sense to me, I have left a couple of minor comments inline. 

One more thing - please use `clang-format` to format the code, you can do that via `git-clang-format` as well after `git add`ing the files, that tool could be found in `clang/tools/clang-format` at your `llvm-project` checkout.

File: lib/Sema/MiscDiagnostics.cpp [hunk 1]
```diff
@@ -2893,7 +2893,20 @@ VarDeclUsageChecker::~VarDeclUsageChecker() {
...
                     noParens = isIsTest = true;
                   }
-                  
+                  // In cases where the value is optional, the cast expr is
+                  // wrapped inside OptionalEvaluationExpr. Unwrap it to get
+                  // ConditionalCheckedCastExpr.
+                  if (auto oeExpr =
+                          dyn_cast<OptionalEvaluationExpr>(initExpr)) {
+                    if (auto ccExpr = dyn_cast<ConditionalCheckedCastExpr>(
+                            oeExpr->getSubExpr())) {
```
[the remark this answers is not in the document]
Comment (mininny) [reply lib/Sema/MiscDiagnostics.cpp]: Thanks for the comment :) I modified the code to use `getValueProvidingExpr`to cover try expressions as well, and added necessary tests!

**Comment by mininny** [comment]
> One more thing - please use `clang-format` to format the code, you can do that via `git-clang-format` as well after `git add`ing the files, that tool could be found in `clang/tools/clang-format` at your `llvm-project` checkout.

Thanks! I applied `clang-format` to format the code :)

File: test/decl/var/usage.swift [hunk 2]
```diff
@@ -504,3 +540,9 @@ func testVariablesBoundInPatterns() {
     break
   }
 }
+// Tests fix to SR-1464
```
Comment (LucianoPAlmeida) [inline test/decl/var/usage.swift:543]: Nit: 
```suggestion
// Tests fix to SR-14646
```

**Review by xedin — APPROVED** [review]

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please clean smoke test macOS platform

**Comment by xedin** [comment]
@swift-ci please clean smoke test macOS platform

**Comment by LucianoPAlmeida** [comment]
@swift-ci please clean smoke test macOS platform

**Comment by xedin** [comment]
@swift-ci please smoke test macOS platform

**Comment by xedin** [comment]
@swift-ci please smoke test macOS platform

**Pull request merged**